products/jbrowse-cli/src/commands/text-index.ts

```diff
@@ -121,82 +136,129 @@ export default class TextIndex extends JBrowseCommand {
-        aggregateStream = gff3Stream.pipe(aggregateStream, { end: false })
-
-        // If a stream ends we have two options:
-        //  1) it is the last stream, so end the aggregate stream.
-        //  2) it is not the last stream, so add a '\n' to separate streams for the indexer.
-
-        // eslint-disable-next-line no-loop-func
-        gff3Stream.once('end', () => {
-          if (--numStreamsFlowing === 0) {
-            aggregateStream.end()
-          } else {
-            aggregateStream.push('\n')
-          }
-        })
+        yield* this.indexGff3(config, attributes, outLocation, quiet)
+      }
+    }
+  }
+
+  async *indexGff3(
+    config: Track,
+    attributes: string[],
+    outLocation: string,
+    quiet: boolean,
+  ) {
+    const {
+      adapter: {
+        gffGzLocation: { uri },
+      },
+      trackId,
+    } = config
+
+    // progress bar code was aided by blog post at
+    // https://webomnizz.com/download-a-file-with-progressbar-using-node-js/
+    const progressBar = new SingleBar(
+      {
+        format: '{bar} ' + trackId + ' {percentage}% | ETA: {eta}s',
+      },
+      Presets.shades_classic,
+    )
+
+    let fileDataStream
+    let totalBytes = 0
+    let receivedBytes = 0
+    if (this.isURL(uri)) {
+      fileDataStream = await this.createRemoteStream(uri)
+      totalBytes = +(fileDataStream.headers['content-length'] || 0)
+    } else {
+      const filename = path.join(outLocation, uri)
+      totalBytes = fs.statSync(filename).size
+      fileDataStream = fs.createReadStream(filename)
+    }
+
+    if (!quiet) {
+      progressBar.start(totalBytes, 0)
+    }
+
+    fileDataStream.on('data', chunk => {
+      receivedBytes += chunk.length
+      progressBar.update(receivedBytes)
+    })
+
+    const gzStream = uri.endsWith('.gz')
+      ? fileDataStream.pipe(createGunzip())
+      : fileDataStream
+
+    const rl = readline.createInterface({
+      input: gzStream,
+    })
+
+    for await (const line of rl) {
+      if (line.startsWith('#')) {
+        continue
+      } else if (line.startsWith('>')) {
+        break
+      }
+
+      const [seq_id, , type, start, end, , , , col9] = line.split('\t')
+      const locStr = `${seq_id}:${start}..${end}`
+
+      if (type !== 'exon' && type !== 'CDS') {
+        const col9attrs = col9.split(';')
+        const name = col9attrs
+          .find(f => f.startsWith('Name'))
+          ?.split('=')[1]
+          .trim()
+        const id = col9attrs
+          .find(f => f.startsWith('ID'))
+          ?.split('=')[1]
+          .trim()
+        const attrs = attributes.map(attr =>
+          col9attrs
+            .find(f => f.startsWith(attr))
+            ?.split('=')[1]
+            .trim(),
+        )
+        if (name || id) {
+          const record = JSON.stringify([locStr, trackId, name, id])
+          const buff = Buffer.from(record).toString('base64')
+          yield `${buff} ${[...new Set(attrs)].join(' ')}\n`
```

Comment:
I don't think we want to have completely fixed data in the records that are returned. Eventually we are going to want to know *why* a record matched, and if the actual data field that matched the query isn't returned from the index lookup, we won't be able to know that unless we go and laboriously fetch the actual feature from the associated data store with a separate query, which we probably would want to avoid.

I think we need to put meta.json back in so we can get more data from the result records

Reply:
I'm am not sure I agree with this need, or at the very least, i'd like to see the plan to restore it. This will bloat the ix files and be more complicated to maintain if not careful

Reply:
See item (8) in original post here for some ideas